### PR TITLE
Refactor E2E tests to separate test cases

### PR DESCRIPTION
**Pull request by periklis**
This PR addresses the fragility of our E2E tests for elasticsearch cluster reconciliation. The present state consists of multiple E2E test cases glued together using the same elasticsearch cluster. This approach yields breaks test isolation. The new state isolates each test case by:
* Providing cleanup after each test case
* Using independent CR instances (e.g. `elasticsearch-$UUID`) and cluster deployments (e.g. `elasticsearch-$UUID-{cd,cdm}-$nodeUUID-...`)
* Usinge independent CR secrets and unique certificates per test case
* Continues to run test cases sequentially to keep resource usage on CI to the old levels. The preference is one cluster at a time thus one test case at a time.

/cc @ewolinetz

### CODE REVIEW

File: test/e2e-olm/elasticsearch_test.go [hunk 1]
```diff
@@ -1,159 +1,240 @@
...
+
+	ctx := test.NewTestCtx(t)
 	namespace, err := ctx.GetNamespace()
 	if err != nil {
-		return fmt.Errorf("Could not get namespace: %v", err)
+		t.Fatal(err)
+	}
+	t.Logf("Found namespace: %v", namespace)
+
+	if err = createElasticsearchSecret(t, f, ctx); err != nil {
```
Comment (ewolinetz) [inline test/e2e-olm/elasticsearch_test.go:39]: this shouldn't change for each test, would it make sense to break this out into some sort of "setup" step? then for a cert full cluster restart we could regenerate as part of that test...

File: test/e2e-olm/elasticsearch_test.go [hunk 2]
```diff
@@ -1,159 +1,240 @@
...
+		t.Fatal(err)
 	}
 
 	dataUUID := utils.GenerateUUID()
 	t.Logf("Using GenUUID for data nodes: %v", dataUUID)
 
+	// Create CR with two nodes sharing client, data and master roles
```
Comment (ewolinetz) [inline test/e2e-olm/elasticsearch_test.go:46]: comment is misleading

File: test/e2e-olm/elasticsearch_test.go [hunk 3]
```diff
@@ -1,159 +1,240 @@
...
 	registerSchemes(t)
-	t.Run("elasticsearch-group", func(t *testing.T) {
-		t.Run("Cluster", ElasticsearchCluster)
-	})
+	t.Run("Single node", singleNodeTest)
+	t.Run("Multiple nodes", multipleNodesTest)
+	t.Run("Multiple nodes with a single non-data node", multipleNodesWithNonDataNodeTest)
+	t.Run("Full cluster redeploy", fullClusterRedeployTest)
+	t.Run("Rolling restart", rollingRestartTest)
+	t.Run("Invalid master count", invalidMasterCountTest)
```
Comment (ewolinetz) [inline test/e2e-olm/elasticsearch_test.go:26]: can we also have a scale up test in here? esp a test when we go from 3 -> 4+ nodese

File: test/e2e-olm/elasticsearch_test.go [hunk 4]
```diff
@@ -256,63 +350,52 @@ func elasticsearchFullClusterTest(t *testing.T, f *framework.Framework, ctx *fra
...
-	}
-
-	// ensure all prior nodes are ready again
-	err = e2eutil.WaitForDeployment(t, f.KubeClient, namespace, fmt.Sprintf("elasticsearch-cdm-%v-1", dataUUID), 1, retryInterval, timeout)
-	if err != nil {
-		return fmt.Errorf("timed out waiting for Deployment %v: %v", fmt.Sprintf("elasticsearch-cdm-%v-1", dataUUID), err)
-	}
-
-	err = e2eutil.WaitForDeployment(t, f.KubeClient, namespace, fmt.Sprintf("elasticsearch-cdm-%v-2", dataUUID), 1, retryInterval, timeout)
+	t.Log("Waiting for redeployment after secret update")
```
Comment (ewolinetz) [inline test/e2e-olm/elasticsearch_test.go:353]: misleading comment

File: test/e2e-olm/utils.go [hunk 5]
```diff
@@ -46,22 +46,22 @@ func registerSchemes(t *testing.T) {
...
+	cpuValue := resource.MustParse("200m")
+	memValue := resource.MustParse("1Gi")
 
 	esDataNode := elasticsearch.ElasticsearchNode{
 		Roles: []elasticsearch.ElasticsearchNodeRole{
 			elasticsearch.ElasticsearchRoleClient,
 			elasticsearch.ElasticsearchRoleData,
 			elasticsearch.ElasticsearchRoleMaster,
 		},
-		NodeCount: int32(1),
```
Comment (ewolinetz) [inline test/e2e-olm/utils.go:64]: the reason this is `1` here is because that is how CLO generates the ES CR

Comment (periklis) [reply test/e2e-olm/utils.go:64]: Taking into account that this is only a creator function. I left the decision to the calling side to use the default or not, e.g. for a multiple nodes cluster I want to start with two nodes already.

File: test/e2e-olm/elasticsearch_test.go (hunk 1 above)
Comment (periklis) [reply test/e2e-olm/elasticsearch_test.go:39]: All calls to `createElasticsearchSecret` register a cleanup routing for the secret. In addition each testcase runs `ctx.Cleanup()` at the end. Thus creating the secret is part of each test case. Since this only creates a secret resource w/o re-generating the certs it is ok to re-create on each test. This ensures that each test is independent.

File: test/e2e-olm/elasticsearch_test.go [hunk 6]
```diff
@@ -1,159 +1,302 @@
...
 	if err != nil {
-		return fmt.Errorf("Could not get namespace: %v", err)
+		t.Fatal(err)
+	}
+	t.Logf("Found namespace: %v", namespace)
+
+	esUUID := utils.GenerateUUID()
+	t.Logf("Using UUID for elasticsearch CR: %v", esUUID)
+
+	if err = createElasticsearchSecret(t, f, ctx, esUUID); err != nil {
```
Comment (ewolinetz) [inline test/e2e-olm/elasticsearch_test.go:44]: since we aren't regenerating our certs at all, is it necessary to create the secret for each test?

File: test/e2e-olm/elasticsearch_test.go [hunk 7]
```diff
@@ -1,159 +1,302 @@
...
+	if err != nil {
+		t.Fatal(err)
+	}
+	t.Logf("Found namespace: %v", namespace)
+
+	esUUID := utils.GenerateUUID()
+	t.Logf("Using UUID for elasticsearch CR: %v", esUUID)
+
+	if err = createElasticsearchSecret(t, f, ctx, esUUID); err != nil {
+		t.Fatal(err)
```
Comment (ewolinetz) [inline test/e2e-olm/elasticsearch_test.go:125]: this seems to all be duplicated from test to test, can we move it into a `setup()` function of some sort?

File: test/e2e-olm/utils.go [hunk 8]
```diff
@@ -1,232 +1,301 @@
...
 
 	}
 
+	if err := generateCertificates(t, namespace, uuid); err != nil {
+		return err
+	}
+
 	elasticsearchSecret := utils.Secret(
-		elasticsearchCRName,
+		elasticsearchNameFor(uuid),
```
Comment (ewolinetz) [inline test/e2e-olm/utils.go:178]: the secret name needs to match the CR name -- it has nothing to do with the UUID

File: test/e2e-olm/elasticsearch_test.go [hunk 9]
```diff
@@ -1,159 +1,302 @@
...
+	cr.Spec.RedundancyPolicy = loggingv1.SingleRedundancy
+
+	t.Logf("Updating redundancy policy to %v", cr.Spec.RedundancyPolicy)
+	if err := updateElasticsearchSpec(t, f, cr); err != nil {
+		t.Fatalf("could not update elasticsearch CR to be SingleRedundancy: %v", err)
+	}
 
 	// Update the secret to force a full cluster redeploy
-	err = updateElasticsearchSecret(t, f, ctx)
+	err = updateElasticsearchSecret(t, f, ctx, esUUID)
```
Comment (ewolinetz) [inline test/e2e-olm/elasticsearch_test.go:290]: why are we passing the UUID for this? the secret expected is the CR name

File: test/e2e-olm/elasticsearch_test.go [hunk 10]
```diff
@@ -176,75 +319,91 @@ func elasticsearchFullClusterTest(t *testing.T, f *framework.Framework, ctx *fra
 	//	return fmt.Errorf("Timed out waiting for full cluster restart to complete")
 	//}
 
-	// ensure all prior nodes are ready again
-	err = e2eutil.WaitForDeployment(t, f.KubeClient, namespace, fmt.Sprintf("elasticsearch-cdm-%v-1", dataUUID), 1, retryInterval, timeout)
+	t.Log("Waiting for redeployment after secret update")
```
Comment (ewolinetz) [inline test/e2e-olm/elasticsearch_test.go:322]: just a side comment, but my concern with this test (as with the original one) is we don't know that the pods have actually been redeployed or if we just move on because they're still in the same state as before.

File: test/e2e-olm/elasticsearch_test.go [hunk 11]
```diff
@@ -256,63 +415,69 @@ func elasticsearchFullClusterTest(t *testing.T, f *framework.Framework, ctx *fra
...
-	}
+	t.Log("Waiting for restart after resourse requests/limits update")
 
-	// ensure all prior nodes are ready again
-	err = e2eutil.WaitForDeployment(t, f.KubeClient, namespace, fmt.Sprintf("elasticsearch-cdm-%v-1", dataUUID), 1, retryInterval, timeout)
-	if err != nil {
-		return fmt.Errorf("timed out waiting for Deployment %v: %v", fmt.Sprintf("elasticsearch-cdm-%v-1", dataUUID), err)
-	}
+	// Increase restart timeout on full cluster redeploy until min masters available
+	restartTimeout := time.Second * 600
```
Comment (ewolinetz) [inline test/e2e-olm/elasticsearch_test.go:421]: similar concern here... we could also make sure the pod spec has been updated?

File: test/e2e-olm/elasticsearch_test.go (hunk 10 above)
Comment (periklis) [reply test/e2e-olm/elasticsearch_test.go:323]: We decided not to wait for pods here.

File: test/e2e-olm/elasticsearch_test.go (hunk 11 above)
Comment (periklis) [reply test/e2e-olm/elasticsearch_test.go:423]: As discussed elsewhere, the last acknowledgement was to wait for available pods that include the changed spec. Testing needs to wait for each deployment to get ready vs. available (`status.ReadyReplicas` vs. `AvailableReplicas`). However, the scale down an up of the cluster get's stuck to the point of no-recovery. One can observe both pods (in a two nodes cluster) to fight against each other and eventually crash loop due to missing shards on the security index:

e.g. pod 1:
```
[2020-07-07T10:51:47,832][WARN ][r.suppressed             ] [elasticsearch-ied234ea-cdm-mh917ikj-1] path: /.security/security/roles, params: {index=.security, id=roles, type=security}
org.elasticsearch.action.NoShardAvailableActionException: No shard available for [get [.security][security][roles]: routing [null]]
```

e.g. pod 2:
```
[2020-07-07T10:51:44,862][WARN ][r.suppressed             ] [elasticsearch-ied234ea-cdm-mh917ikj-2] path: /.security/security/roles, params: {index=.security, id=roles, type=security}
org.elasticsearch.action.NoShardAvailableActionException: No shard available for [get [.security][security][roles]: routing [null]] 
```

In summary this needs more time of investigation. Thus I will split the extra spec chenk to a separate PR, because this PR already adds the required reliability to unblock other gated PRs (e.g. #400, #409, #410, #414).

**Comment by periklis** [comment]
@ewolinetz PTAL a final look on this PR to land this ASAP and unblock other gated PRs. Consider my [comment](https://github.com/openshift/elasticsearch-operator/pull/412#discussion_r450781939) above.
